starknet_py/net/signer/ledger_signer.py

```diff
@@ -0,0 +1,169 @@
+from typing import List
+
+from bip_utils import Bip32KeyIndex, Bip32Path, Bip32Utils
+from ledgerwallet.client import LedgerClient
+
+from starknet_py.constants import (
+    EIP_2645_PATH_LENGTH,
+    EIP_2645_PURPOSE,
+    PUBLIC_KEY_LENGTH,
+    SIGNATURE_LENGTH,
+    STARKNET_CLA,
+    VERSION_LENGTH,
+)
+from starknet_py.net.models import AccountTransaction
+from starknet_py.net.models.chains import ChainId
+from starknet_py.net.signer import BaseSigner
+from starknet_py.utils.typed_data import TypedData
+
+
+class LedgerStarknetApp:
+    def __init__(self):
+        self.client: LedgerClient = LedgerClient(cla=STARKNET_CLA)
+
+    @property
+    def version(self) -> str:
+        """
+        Get the Ledger app version.
+
+        :return: Version string.
+        """
+        response = self.client.apdu_exchange(ins=0)
+        if len(response) != VERSION_LENGTH:
+            raise ValueError(
+                f"Unexpected response length (expected: {VERSION_LENGTH}, actual: {len(response)}"
+            )
+        major, minor, patch = list(response)
+        return f"{major}.{minor}.{patch}"
+
+    def get_public_key(
+        self, derivation_path: Bip32Path, device_confirmation: bool = False
+    ) -> int:
+        """
+        Get public key for the given derivation path.
+
+        :param derivation_path: Derivation path of the account.
+        :param device_confirmation: Whether to display confirmation on the device for extra security.
+        :return: Public key.
+        """
+
+        data = _derivation_path_to_bytes(derivation_path)
+        response = self.client.apdu_exchange(
+            ins=1,
+            data=data,
+            p1=int(device_confirmation),
+            p2=0,
+        )
+
+        if len(response) != PUBLIC_KEY_LENGTH:
+            raise ValueError(
+                f"Unexpected response length (expected: {PUBLIC_KEY_LENGTH}, actual: {len(response)}"
+            )
+
+        public_key = int.from_bytes(response[1:33], byteorder="big")
+        return public_key
+
+    def sign_hash(self, derivation_path: Bip32Path, hash_val: int) -> List[int]:
+        """
+        Request a signature for a raw hash with the given derivation path.
+        Currently, the Ledger app only supports blind signing raw hashes.
+
+        :param derivation_path: Derivation path of the account.
+        :param hash_val: Hash to sign.
+        :return: Signature as a list of two integers.
+        """
+
+        data = _derivation_path_to_bytes(derivation_path)
+        self.client.apdu_exchange(
+            ins=0,
+            data=data,
+            p1=0,
+            p2=0,
+        )
+
+        # for some reason the Ledger app expects the data to be left shifted by 4 bits
```

Comment:
Unfortunately I could't find any official information about this :/. [APDU spec](https://github.com/LedgerHQ/app-starknet/blob/develop/docs/apdu.md#command-2) seems like a place in which this should be documented. Initially, without shifting, the test for deploy account and transfer failed because of invalid signature. After shifting, the signature was correct and it passed

In [rust implementation](https://github.com/xJonathanLEI/starknet-rs/blob/f6d339c6b897fb38c839485608ca2fe374a6275d/starknet-signers/src/ledger.rs#L318) there was a same remark.